// authorization_command/index.js
module.exports = class authorization_command {
    
    /**
     * Constructor authorization_command object
     * @constructor
     * @param {string} dataDir - The path to the directory which the data on authorization for the commands.
     * @param {string} dataFileName - The name of the file which hold the data on authorization for the command. Default value is 'AUTHORIZATION'
     * @param {string} settingsDir - The path to the directory which the data on authorize setting. Default value is the value of settingsDir
     * @param {string} settingsFileName - The name of the file which hold the data on authorize setting. Default value is 'AUTHORIZE_ROLE'
     */
    constructor(dataDir, dataFileName, settingsDir, settingsFileName) {
        if (dataDir === '') {
            throw new Error('authorization_command didn\'t get data directory path');
        }
        this.dataDir = dataDir;
        if (dataFileName === '' || dataFileName === undefined) {
            this.dataFileName = 'AUTHORIZATION';
        } else {
            this.dataFileName = dataFileName;
        }

        if (settingsDir === '' || settingsDir === undefined) {
            this.settingsDir = dataDir;
        } else {
            this.settingsDir = settingsDir;
        }

        if (settingsFileName === '' || settingsFileName === undefined) {
            this.settingsFileName = 'AUTHORIZE_ROLE';
        } else {
            this.settingsFileName = settingsFileName;
        }

        this.authArr = load_auth_data(this.dataDir, this.dataFileName);
        this.authSettingsArr = load_auth_settings(this.settingsDir, this.settingsFileName);
    }
    
    /**
     * Return true or false for, if the member is allowed to use the command. Used for commands which are called through DM
     * @param {GuildMember} member - The guild member which want to use the command
     * @param {string} id - id of the command
     * @returns {boolean} - Is the member is allowed to use the command
     */
    auth_dm_command(member, id) {
        if (member === undefined) throw new Error('auth_dm_command didn\'t get a member');
        if (id === undefined) throw new Error('auth_dm_command didn\'t get a id');
        console.log('auth_dm_command');
    }
    
    /**
     * Return true or false for, if the member is allowed to use the command. Used for commands which are called through the server
     * @param {GuildMember} member - The guild member which want to use the command
     * @param {string} id - id of the command
     * @returns {boolean} - Is the member is allowed to use the command
     */
    auth_text_command(member, id) {
        if (member === undefined) throw new Error('auth_text_command didn\'t get a member');
        if (id === undefined) throw new Error('auth_text_command didn\'t get a id');
        const constants = require('consts');
        if (this.authArr[id].type === constants.AUTHORIZATION_ROLE) { //If the command with auth by the member role
            return auth_text_command_by_role(member, this.authArr[id]);
        } else if (this.authArr[id].type === constants.AUTHORIZATION_SETTINGS) {//If the command with auth by the member settings
            return auth_text_command_by_settings(member, this.authArr[id], this.authSettingsArr);
        }
        return true;
    }
    
    /**
     * Update \ create a command authorization
     * @param {string} id - Id of the command
     * @param {string} type - The type of the authorization. For examples: 'role' 'settings'
     * @param {Array} authGroup - The group which is authorize to use the command. For example, if type is 'role', then the array will hold all the roles which are allowed to use the command
     * @returns {void}
     */
    auth_update_command(id, type, authGroup) {

    }
    
    /**
     * Remove the command from authorization
     * @param {string} id - Id of the command
     * @returns {void}
     */
    auth_remove_command(id) {

    }
    
    /**
     * Create a new authorize setting
     * @param {string} setting - The name of the new setting
     * @returns {void}
     */
    auth_create_setting_command(setting) {

    }
    
    /**
     * Remove authorize setting
     * @param {string} setting - The name of the setting which will be removed
     * @returns {void}
     */
    auth_remove_setting_command(setting) {

    }
    
    /**
     * Adding user to authorize setting
     * @param {string} setting - The name of the setting which will be removed
     * @param {GuildMember} member - Member which will be added to the authorize setting
     * @returns {void}
     */
    auth_add_user_to_settings_command(setting, member) {

    }
};

/**
 * Return the authorization data from the file, without using the cache, from the given directory and file name
 * @param {string} dataDir - The directory where the file is
 * @param {string} dataFileName - The name of the file which hold the data for the settings of the auth
 * @returns {object} - The authorization data from the file
 */
function load_auth_data(dataDir, dataFileName) {
    const filePath = dataDir + '\\' + dataFileName;

    delete require.cache[require.resolve(filePath)]; //delete cache
    return require(filePath);
}

/**
 * Return the auth settings data from the file, without using the cache, from the given directory and file name
 * @param {string} settingsDir - The directory where the file is
 * @param {string} settingsFileName - The name of the file which hold the data for the settings of the auth
 * @returns {object} - The authorization settings from the file
 */
function load_auth_settings(settingsDir, settingsFileName) {
    const filePath = settingsDir + '\\' + settingsFileName;

    delete require.cache[require.resolve(filePath)]; //delete cache
    return require(filePath);
}

/**
 * Return true or false for, if the member is allowed to use the command based on the member roles in the guild
 * @param {GuildMember} member - The guild member which want to use the command
 * @param {Object} authData - Authorization data on a spesific command
 * @returns {boolean} - Is the member is allowed to use the command
 */
function auth_text_command_by_role(member, authData) {
    //filter that will only take roles which the member have and the role exist in authData
    const filteredRoles = member.roles.filter(function (role) { return authData.roles.includes(role.name); });
    if (filteredRoles.size > 0) { return true; } //If better then 0, the member have the role to use the command
    else { return false; } //The member doeesn't have the auth to use the command
}

/**
 * Return true or false for, if the member is allowed to use the command based on the member roles\settings in the bot. Root able to do all
 * @param {GuildMember} member - The guild member which want to use the command
 * @param {Object} authData - Authorization data on a spesific command
 * @param {Array} authSettings - Authorization settings. The bot roles group
 * @returns {boolean} - Is the member is allowed to use the command
 */
function auth_text_command_by_settings(member, authData, authSettings) {
    const constants = require('consts');

    //If member is root, then let the user use the command without checking
    if (authSettings[constants.AUTHORIZATION_SETTINGS_ROOT].includes(member.user.username)) {
        return true;
    }

    //foreach settings the commands has:    
    //For given setting for the command, will check, if member has it. If found, return true, else continue search until the end
    let memberAllowed = false;
    authData.settings.forEach(function (setting) {
        if (authSettings[setting].includes(member.user.username)) {
            memberAllowed = true; //If it's find that the member belong to one of the settings which are allowed to use the command, will turn to true
        }
    });

    return memberAllowed;
}
