// authorization_command/index.js
module.exports = class authorization_command {
  
	constructor(dataDir, dataFileName) {
        if(dataDir === '') {
        throw new Error('authorization_command didn\'t get data directory path');
        }
        this.dataDir = dataDir;
        if(dataFileName === '' || dataFileName === undefined) {
        this.dataFileName = 'AUTHORIZATION';
        } else {
        this.dataFileName = dataFileName;
        }

        this.authArr = load_auth_data(this.dataDir, this.dataFileName);
	}
  
	auth_dm_command() {
        console.log('auth_dm_command');
    }

    auth_text_command() {
        console.log('auth_text_command');
        auth_text_command_by_role();
        auth_text_command_by_settings();

        console.log(this.authArr);
        console.log(this.authArr.test_server_command);
    }
};

function load_auth_data(dataDir, dataFileName) {
    const filePath = dataDir + '\\'+ dataFileName;

    delete require.cache[require.resolve(filePath)];
    return require(filePath);
}

function auth_text_command_by_role() {
    console.log('auth_text_command_by_role');
}

function auth_text_command_by_settings() {
    console.log('auth_text_command_by_settings');
}
